modules/internal/ArrayViewRankChange.chpl

```diff
@@ -113,6 +113,10 @@ module ArrayViewRankChange {
     override proc dsiIsLayout() param {
       return downDistInst.dsiIsLayout();
     }
+
+    proc dsiEqualDMaps(that: ArrayViewRankChangeDist(?)) {
+      return this.downDist.dsiEqualDMaps(that.downDist);
```

Comment:
I'm trying to think whether a check for equal domain maps should also ensure that the collapsed dimension and/or index of the rank change are equal as well, where my intuition is "yes" (but my skepticism is that I'm never entirely clear what this routine is used for).  For example, I'd expect `A[i, 1..n].dist` to be != to `A[1..n, j].dist` and to `A[j, 1..n]` (and maybe even to `A[i, j..k].dist`?) but I think that this implementation would make all of these equal since it's only comparing A's domain map?

Reply:
I was thinking that those should be equal? My thinking is that the rank change domain is distributed the same way as its base domain is, such that any index you pick within the view should be on the same locale as the base domain's index.

Reply:
This gets to my uncertainty about how these routines are used / what they are intended to mean.  Is it "are these distributed in an aligned manner?" (which is what I think you're saying?) or "are they equivalent distributions?" (which is what I was expecting / thinking these meant).

As an example, a difference might be that if I were to say:

```chapel
var D2 = {1..n, 1..n} dmapped A.dist;
var D1 = {1..n} dmapped A[i, 1..n].dist;
```

I might be considered to have ended up with aligned distributions (by some definition), but it still doesn't seem like they are equal distributions in the sense of `==`, in that one is a 1D distribution and the other a 2D distribution (and since `==` maps down to this routine for domain maps, that would seem surprising to me).

If I haven't mistaken anything here, the fix might be as simple as introducing some other `isAligned()` query on domain maps.  Though it still would seem a little odd to me that the two domain maps in the example above might be considered aligned when they are of different rank since it raises a question of "which specific indices of D2 are aligned with indices of D1?"

Reply:
I think I am getting it... To convince myself, if we have a 4x4 Block array and create two rank-change views A and B and they are like:

```
 A| B
 A| B
-----
 A| B
 A| B
```

They are unaligned here, but with my initial thinking and implementation, we'll get otherwise. And that we should check whether they are collapsed on the same dimension and index (we can probably relax that a little).

One of the misconceptions that I have and I am now realizing is that if you  get a 1D view from a 2D array, and iterate over the domain of the view, I was expecting that you'd get 2D indices.

So, I was thinking that this check will be used for cases like the following

```chapel
var a: [D] int, b:[D] int;

ref aView = a.[1..n, j];
ref bView = a.[1..n, j];

forall i in bView.domain {
  .... aView[i] ....
}
```

where both views must be created the same way, but now I am seeing that j's can be different.

Reply:
As for the solution to the problem at hand, I was leaning towards fixing the implementation of dsiEqualDMaps here to check for the collapsed dimension and index as well as the downdist's equivalence, but that might make the implementation of that method too strict. That would return not equal for A and B here:

```
BA|  
BA|  
-----
BA|  
BA|  
```

So, maybe the solution for now is to make that strict check within the auto local access implementation, and not implement this function yet (probably capturing the discussion here in a separate issue). What do you think?

Reply:
> One of the misconceptions that I have and I am now realizing is that if you get a 1D view from a 2D array, and iterate over the domain of the view, I was expecting that you'd get 2D indices.

That's right.  Once you've created a rank-change array view, its domain and distribution also act as though they are the lower-dimensional rank, such that `var B: [A[1..n, j].domain] real;` will create a 1D array B aligned with `1..n, j` of A.

Reply:
> That's right. Once you've created a rank-change array view, its domain and distribution also act as though they are the lower-dimensional rank, such that var B: [A[1..n, j].domain] real; will create a 1D array B aligned with 1..n, j of A.

This was OK, actually. My thinking was that the view domain would know that it was 1D, but it would also know where it is in the 2D plane. So, you can iterate over the view domain and access the base array using the indices yielded from the iterator. Which may be powerful or weird, I don't know :)

Reply:
In your diagram:

```
BA|  
BA|  
-----
BA|  
BA|  
```

is the idea that B is made from a [1..n, i] rank change slice and A from a [1..n, j] rank change slice where i != j but they happen to fall on the same locale slice?

If so, I don't think I'd treat these as equal to be on the safe side.  Specifically, this feels a lot like saying that `Block({1..n, 1..n})` and `Block({0..n+1, 0..n+1})` are equal if we're doing an `-nl 1` execution because they both happen to map the same indices to the same locale.  But to me, that's more about alignment than it is about equality, where I'd consider equality to be more like "is guaranteed to behave the same in all executions."  So by adding enough locales, [1..n, i] and [1..n, j] rank change slices could end up targeting different locales and shouldn't be considered equal either.

But again, for optimizations like this, maybe the question we really want to be asking is "are aligned" rather than "are equal"?

Reply:
> but it would also know where it is in the 2D plane.

It does, but...

> So, you can iterate over the view domain and access the base array using the indices yielded from the iterator. Which may be powerful or weird, I don't know :)

...I'd call it weird, in that (continuing the previous example):

```chapel
var B: [A[1..n, j].domain] real;
forall i in B.domain do
  B[i] += 1;
```

would give you "rank mismatch" errors for `B[i]` if `B.domain` yielded 2D indices.

Reply:
> So by adding enough locales, [1..n, i] and [1..n, j] rank change slices could end up targeting different locales and shouldn't be considered equal either.

Good point.

> But again, for optimizations like this, maybe the question we really want to be asking is "are aligned" rather than "are equal"?

Also true.

So, what I am thinking is to implement equality as you suggest and use it for dynamic auto local access checks.
And create an issue for alignment checks for domains/distributions, and once we have it we can use it for auto local access, too.

Reply:
I created https://github.com/chapel-lang/chapel/issues/17069 to capture the discussion about alignment here.